gateway/exporters/statsd/statsd.go

```diff
@@ -0,0 +1,225 @@
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/cactus/go-statsd-client/v5/statsd"
+	"github.com/openconfig/gnmi-gateway/gateway/configuration"
+	"github.com/openconfig/gnmi-gateway/gateway/connections"
+	"github.com/openconfig/gnmi-gateway/gateway/exporters"
+	"github.com/openconfig/gnmi-gateway/gateway/utils"
+	"github.com/openconfig/gnmi/ctree"
+	gnmipb "github.com/openconfig/gnmi/proto/gnmi"
+)
+
+const Name = "statsd"
+
+var _ exporters.Exporter = new(StatsdExporter)
+
+type Metric struct {
+	Account     string            `json:"Account"`
+	Measurement string            `json:"Metric"`
+	Namespace   string            `json:"Namespace"`
+	Dims        map[string]string `json:"Dims"`
+	Value       interface{}       `json:"-"`
+}
+
+type Point struct {
+	Tags   map[string]string
+	Fields map[string]interface{}
+}
+
+func init() {
+	exporters.Register(Name, NewStatsdExporter)
+}
+
+func NewStatsdExporter(config *configuration.GatewayConfig) exporters.Exporter {
+	return &StatsdExporter{
+		config: config,
+	}
+}
+
+type StatsdExporter struct {
+	config  *configuration.GatewayConfig
+	connMgr *connections.ConnectionManager
+	client  statsd.Statter
+}
+
+func (e *StatsdExporter) Name() string {
+	return Name
+}
+
+func (e *StatsdExporter) Export(leaf *ctree.Leaf) {
+	notification := leaf.Value().(*gnmipb.Notification)
+
+	for _, update := range notification.Update {
+		point := Point{
+			Fields: make(map[string]interface{}),
+		}
+
+		metric := Metric{}
+
+		timestamp := time.Unix(0, notification.Timestamp)
+		beforeLimit := (time.Now()).Add(-30 * time.Minute)
+		afterLimit := (time.Now()).Add(4 * time.Minute)
+
+		if timestamp.Before(beforeLimit) || timestamp.After(afterLimit) {
+			return
+		}
+
+		var name string
+
+		value, valid := utils.GetValues(update.Val)
+
+		if !valid {
+			continue
+		}
+
+		metric.Value = value
+
+		elems, keys, err := extractPrefixAndPathKeys(notification.GetPrefix(), update.GetPath())
+		if err != nil {
+			e.config.Log.Info().Msg(fmt.Sprintf("Failed to extract path or keys: %s", err))
+		}
+
+		name, elems = elems[len(elems)-1], elems[:len(elems)-1]
+
+		if metric.Measurement == "" {
+			measurementPath := strings.Join(elems, "/")
+
+			p := notification.GetPrefix()
+			target := p.GetTarget()
+			origin := p.GetOrigin()
+
+			keys["path"] = measurementPath
+			if origin != "" {
+				keys["origin"] = origin
+			}
+
+			if target != "" {
+				keys["target"] = target
+			}
+
+			metric.Measurement = pathToMetricName(measurementPath)
+
+			point.Tags = keys
+
+			targetConfig, found := (*e.connMgr).GetTargetConfig(point.Tags["target"])
+
+			if found {
+				for _, fieldName := range e.config.ExporterMetadataAllowlist {
+					fieldVal, exists := targetConfig.Meta[fieldName]
+					if exists {
+						point.Tags[fieldName] = fieldVal
+					} else {
+						e.config.Log.Error().Msg("Field " + fieldName + " is not set for device " + point.Tags["target"] + " but is listed in the exporter's Metadata allow-list")
```

Comment:
The allow list may at some point contain optional fields, in which case we shouldn't log an error. I guess it's ok for now.